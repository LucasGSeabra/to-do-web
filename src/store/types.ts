import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { RootState, AppDispatch } from './store';

export interface todo {
  _id: string
  done: boolean
  description: string
  createdAt: string
}

export interface todoState {
  description: string
  list: todo[]
}

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
