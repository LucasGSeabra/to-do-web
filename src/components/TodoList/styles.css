.list th {
    border-top: none !important;
}

.list .table-actions {
    width: 12rem !important;
}
