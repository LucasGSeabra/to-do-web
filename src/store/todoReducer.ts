import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';
import { todoState, todo } from './types';

const URL = 'https://glacial-castle-07790.herokuapp.com/api/todos';

const initialState: todoState = {
  description: '',
  list: [],
};

export const fetchTodos = createAsyncThunk('fetchTodos', async (parameter: string) => {
  const search = parameter ? `&description__regex=/${parameter}/` : '';
  const response = await axios.get(`${URL}?sort=-createdAt${search}`);
  return response.data;
});

export const addTodos = createAsyncThunk('addTodos', async (parameter: string) => {
  const description = parameter;
  const response = await axios.post(URL, { description });
  return response.data;
});

export const changeStatus = createAsyncThunk('changeStatus', async (parameter: todo) => {
  const item = parameter;
  const response = await axios.put(`${URL}/${item._id}`, { ...item, done: !item.done });
  return response.data;
});

export const removeTodos = createAsyncThunk('removeTodos', async (parameter: string) => {
  const id = parameter;
  await axios.delete(`${URL}/${id}`);
  return id;
});

export const todoReducer = createSlice({
  name: 'todo',
  initialState,
  reducers: {
    descriptionChanged(state, action) {
      return { ...state, description: action.payload };
    },
    descriptionCleared(state) {
      return { ...state, description: '' };
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchTodos.fulfilled, (state, action) => ({ ...state, list: action.payload }))
      .addCase(addTodos.fulfilled, (state, action) => {
        const newList = state.list;
        newList.unshift(action.payload);
        state.description = '';
      })
      .addCase(changeStatus.fulfilled, (state, action) => {
        const { _id, done } = action.payload;
        const item = state.list.find((todoItem) => todoItem._id === _id);
        item!.done = done;
      })
      .addCase(removeTodos.fulfilled, (state, action) => {
        const id = action.payload;
        const newList = state.list.filter((todoItem) => todoItem._id !== id);
        return { ...state, list: newList };
      });
  },
});

export const { descriptionChanged, descriptionCleared } = todoReducer.actions;

export default todoReducer.reducer;
