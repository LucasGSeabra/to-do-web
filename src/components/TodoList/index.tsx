import React, { useEffect } from 'react';
import { Table } from 'react-bootstrap';
import { useAppDispatch, useAppSelector } from '../../store/hooks';
import { fetchTodos } from '../../store/todoReducer';
import TodoItem from '../TodoItem';

import './styles.css';

function TodoList() {
  const list = useAppSelector((state) => state.todo.list);
  const description = useAppSelector((state) => state.todo.description);
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(fetchTodos(description));
  }, []);

  return (
    <Table className="list">
      <thead>
        <tr>
          <th>Descricao</th>
          <th className="table-actions">Ações</th>
        </tr>
      </thead>
      <tbody>
        {list.map((todo) => (
          <TodoItem item={todo} key={todo._id} />
        ))}
      </tbody>
    </Table>
  );
}

export default TodoList;
