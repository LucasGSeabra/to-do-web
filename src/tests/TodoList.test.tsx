/* eslint-disable no-undef */
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import TodoList from '../components/TodoList';
import store from './testReducer';

describe('TodoList Component', () => {
  it('should render TodoList correctly', () => {
    const component = renderer.create(
      <Provider store={store}>
        <TodoList />
      </Provider>,
    ).toJSON();
    expect(component).toMatchSnapshot();
  });
});
