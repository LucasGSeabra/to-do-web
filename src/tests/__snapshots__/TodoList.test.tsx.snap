// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList Component should render TodoList correctly 1`] = `
<table
  className="list table"
>
  <thead>
    <tr>
      <th>
        Descricao
      </th>
      <th
        className="table-actions"
      >
        Ações
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        className="marked"
      >
        Teste 1 True
      </td>
      <td>
        <button
          className="btn btn-warning"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-redo fa-w-16 "
            data-icon="redo"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M500.33 0h-47.41a12 12 0 0 0-12 12.57l4 82.76A247.42 247.42 0 0 0 256 8C119.34 8 7.9 119.53 8 256.19 8.1 393.07 119.1 504 256 504a247.1 247.1 0 0 0 166.18-63.91 12 12 0 0 0 .48-17.43l-34-34a12 12 0 0 0-16.38-.55A176 176 0 1 1 402.1 157.8l-101.53-4.87a12 12 0 0 0-12.57 12v47.41a12 12 0 0 0 12 12h200.33a12 12 0 0 0 12-12V12a12 12 0 0 0-12-12z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
        <button
          className="btn btn-danger"
          disabled={false}
          hidden={false}
          onClick={[Function]}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-trash fa-w-14 "
            data-icon="trash"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </td>
    </tr>
    <tr>
      <td
        className=""
      >
        Teste 2 False
      </td>
      <td>
        <button
          className="btn btn-success"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-check fa-w-16 "
            data-icon="check"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
        <button
          className="btn btn-danger"
          disabled={false}
          hidden={true}
          onClick={[Function]}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-trash fa-w-14 "
            data-icon="trash"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
